### ISSUE TITLE
these objects' docstrings are not proper epytext:

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2375]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#2375|
|Type|defect|
|Created|2007-01-08 23:58:12Z|
|Branch|https://github.com/twisted/twisted/tree/proper-epytext-2375|

these objects' docstrings are not proper epytext:
     twisted.python.modules
     twisted.python.modules._ModuleIteratorHelper.__getitem__
     twisted.python.modules._ModuleIteratorHelper.__getitem__
     twisted.python.zshcomp
     twisted.vfs.backends.adhoc.AdhocDirectory.child
     twisted.vfs.backends.adhoc.AdhocDirectory.child
     twisted.vfs.backends.inmem.FakeDirectory.child
     twisted.vfs.backends.inmem.FakeDirectory.child
     twisted.vfs.ivfs.IFileSystemContainer.child
     twisted.vfs.ivfs.IFileSystemContainer.child
     twisted.words.protocols.jabber.client.XMPPClientFactory
     twisted.words.protocols.jabber.client.XMPPClientFactory
     twisted.words.xish.domish.Element
     twisted.web2.dav.xattrprops.xattrPropertyStore
     twisted.web2.dav.resource.DAVPropertyMixIn
     twisted.web2.client.http.EmptyHTTPClientManager.clientPipelining
     twisted.web2.client.http.EmptyHTTPClientManager.clientPipelining
     twisted.web2.client.http.EmptyHTTPClientManager.clientPipelining
     twisted.web2.client.interfaces.IHTTPClientManager.clientPipelining
     twisted.web2.client.interfaces.IHTTPClientManager.clientPipelining
     twisted.web2.client.interfaces.IHTTPClientManager.clientPipelining
     twisted.mail.maildir.StringListMailbox.listMessages
     twisted.mail.maildir.StringListMailbox.getMessage
     twisted.mail.maildir.StringListMailbox.listMessages
     twisted.mail.maildir.StringListMailbox.getMessage
     twisted.mail.pop3.IMailbox.listMessages
     twisted.mail.pop3.IMailbox.getMessage
     twisted.mail.pop3.IMailbox.listMessages
     twisted.mail.pop3.IMailbox.getMessage
     twisted.mail.pop3.Mailbox.listMessages
     twisted.mail.pop3.Mailbox.getMessage
     twisted.mail.pop3.Mailbox.listMessages
     twisted.mail.pop3.Mailbox.getMessage
     twisted.internet.defer.inlineCallbacks
     twisted.internet.defer.inlineCallbacks
     twisted.internet._dumbwin32proc._findShebang
     twisted.internet._dumbwin32proc._findShebang
     twisted.spread.jelly.Jellyable.jellyFor
     twisted.spread.flavors.RemoteCopy.unjellyFor
     twisted.spread.flavors.RemoteCopy.unjellyFor
     twisted.spread.flavors.RemoteCache.unjellyFor
     twisted.spread.flavors.RemoteCache.unjellyFor
     twisted.spread.jelly.Jellyable.jellyFor
     twisted.spread.jelly.Jellyable.jellyFor
     twisted.spread.jelly.Jellyable.jellyFor
     twisted.spread.jelly.Unjellyable.unjellyFor
     twisted.spread.jelly.Unjellyable.unjellyFor
     twisted.spread.banana.Banana.setPrefixLimit
     twisted.spread.banana.Banana.setPrefixLimit

Attachments:

* [proper-epytext-2375.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4d2/4d2b62d9c64119cb4313e92b233c7793c63e6302/5c7c50578e78067be8ee021a6805a673d9304efc.patch) (40515 bytes) - added by thijs on 2008-07-13 04:31:44Z - patch against r24262
* [epy-fix.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4d2/4d2b62d9c64119cb4313e92b233c7793c63e6302/eb5b72e0c243c02bfb5f4d708eb831abf0d2a07d.py) (953 bytes) - added by thijs on 2008-07-17 11:08:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2375 2375
type__defect defect
reporter__radix radix
priority__normal normal
milestone__ 
branch__branches_proper_epytext_2375 branches/proper-epytext-2375
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1168300692000000 1168300692000000
changetime__1224938954000000 1224938954000000
version__None None
owner__ 
cc__mwh cc__thijs cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml set owner to @radix|
|-|-|

It would be helpful if you could show how you generated this list, and how one might verify that a given object has a valid epytext docstring. That way, I can fix this bug :)

For bonus points, make pyflakes report invalid epytext.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_2"></a>|@mwhudson commented|
|-|-|

You get this list when you run PyDoctor from SVN head against Twisted SVN HEAD.

I've been talking ([#2272](https://github.com/twisted/twisted/issues/2272)) about setting up a pydoctor buildbot to catch this sort of thing closer to source, but been far too busy to do it.  Integrating it into pyflakes shouldn't be very hard and would be almost as good.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

List for current trunk (r24264) where 28 objects' docstrings are not proper epytext:
```
    twisted.mail.test.test_pop3
    twisted.python.test.test_util.PasswordTestingProcessProtocol
    twisted.conch.test.test_conch.ConchTestForwardingProcess.__init__
    twisted.conch.test.test_helper.BufferTestCase.test_linefeed
    twisted.conch.test.test_helper.BufferTestCase.test_newline
    twisted.conch.test.test_insults.ServerProtocolOutputTests.test_nextLine
    twisted.flow.test.test_flow.slowlist
    twisted.flow.test.test_flow.producer
    twisted.flow.test.test_flow.consumer
    twisted.flow.test.test_flow.badgen
    twisted.flow.test.test_flow.buildlist
    twisted.flow.test.test_flow.testconcur
    twisted.flow.test.test_flow.echoServer
    twisted.flow.test.test_flow.echoClient
    twisted.internet.test.test_iocp.IOCPReactorTestCase.test_stopStartReading
    twisted.python.test.test_release.ProjectTest.makeProject
    twisted.python.test.test_release.DistributionBuilderTests.test_subProjectLayout
    twisted.python.test.test_release.DistributionBuilderTests.test_coreProjectLayout
    twisted.test.test_amp.BinaryProtocolTests.test_receiveBoxStateMachine
    twisted.test.test_ftp.FTPClientTestCase.test_lostRETR
    twisted.test.test_internet.TimeTestCase.test_seconds
    twisted.test.test_process.MockProcessTestCase.test_mockForkInParentGarbageCollectorEnabled
    twisted.test.test_process.MockProcessTestCase.test_mockForkInParentGarbageCollectorDisabled
    twisted.test.test_reflect.LookupsTestCase.test_invalidNames
    twisted.trial.test.test_reporter.PyunitTestNames.test_minimalReporter
    twisted.web.test.test_xml.TestBrokenHTML
    twisted.web2.test.test_vhost.TestVhost.testNameVirtualHost
    twisted.web2.test.test_wsgi.TestContainer.flushErrors
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

The [[proper-epytext-2375.patch](None/proper-epytext-2375.patch) attached patch] fixes all epydoc issues found by pydoctor.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

I don't understand: exarkun, these errors should be reporter by the documentation buildbot slave, no?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

They seem all to be in tests; I think pydoctor ignores the docstrings of tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

So, maybe we want to fix these, but we should get automated coverage so that they remain correct.  This means changing the pydoctor model, I think, or extending the buildbot API generation step to check them in some other way.

On the other hand, if we don't actually include test docstrings in the generated API docs, then it doesn't really matter if the markup is correct or not.  Perhaps instead we should stop using epytext in test docstrings?

Actually I don't like that idea.  I think we should generate API docs for the tests - perhaps separately from the rest of the docs (but perhaps not)?  These could be useful for people (for example, people trying to write their own tests for Twisted-using code).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [exarkun](#note_8):
> They seem all to be in tests; I think pydoctor ignores the docstrings of tests.
Most of them are except for:

 - twisted/protocols/ftp.py
 - twisted/trial/reporter.py

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to @exarkun<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_9):
> So, maybe we want to fix these, but we should get automated coverage so that they remain correct.  This means changing the pydoctor model, I think, or extending the buildbot API generation step to check them in some other way.

Yes, this probably means extending pydoctor and/or the buildbot API generation step. I attached a crappy script that reads a file containing the list of objects reported as broken by pydoctor, and then runs epydoc with warnings enabled on every module, displaying the errors. The script is absolutely not a fix or solution for pydoctor but it worked for me.

> On the other hand, if we don't actually include test docstrings in the generated API docs, then it doesn't really matter if the markup is correct or not.  Perhaps instead we should stop using epytext in test docstrings? Actually I don't like that idea.  

I don't like that idea either.

> I think we should generate API docs for the tests - perhaps separately from the rest of the docs (but perhaps not)?  

I think it's good to use epytext in the tests, so people can generate pretty docs for them if they need docs for the tests, but I don't see why the majority of the users would need API docs for the tests. So generating API docs for the tests for users isn't that important imo, as long as they can do it themselves. Sacrificing a buildbot and cycles for testing the epydoc strings in the tests seems like overkill.

> These could be useful for people (for example, people trying to write their own tests for Twisted-using code).

Assigning back to exarkun for review, not sure what the current status of this ticket is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

> Sacrificing a buildbot and cycles for testing the epydoc strings in the tests seems like overkill.

This is basically the core of it for me.  If it's not worth making sure it stays right, then I don't feel like it's worth fixing.

So I think we should close this as fixed/wontfix or turn the epy-fix.py into something that can run on buildbot in the documentation builder.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set status to `assigned`|
|-|-|

There are one or two files that don't belong to the tests but contain incorrect epytext, I suggest we fix those and ignore the rest. (that way at least some of all the work I spent to correct it will be merged :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

I'll double-check if my previous comment is still valid and then fix it, otherwise close the ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

(In [25031]) Branching to 'proper-epytext-2375'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [thijs](#note_13):
> There are one or two files that don't belong to the tests but contain incorrect epytext, I suggest we fix those and ignore the rest. (that way at least some of all the work I spent to correct it will be merged :)

Fixed the errors in those 2 files in r25032, putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

The `twisted/protocols/ftp.py` changes look good, and I filed a ticket upstream with pydoctor so that the kind of error represented by `@param commands` will be detected and reported.

Some of the changes to `twisted/trial/reporter.py` look good. I'm not sure about the changes from `*args` to `args` though.  We're probably abusing epytext here, but we do this in a lot of places in Twisted, not just `reporter.py`.  Changing just these two instances doesn't fix the overall problem.  We need to figure out if `*args` will actually do something useful, then we can either keep using it or stop using it everywhere.  (for reference, `grep -E '@param \*' twisted/ -r --include '*.py'`)

I'd suggest dropping the `*args`/`args` change and merging the rest.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra set status to `assigned`|
|-|-|

When I run epydoc in warning mode on reporter.py it throws the following error:

```
+--------------------------------------------------------------------------------------------------------------------------------------------
| File /trunk/twisted/trial/reporter.py, line 399, in twisted.trial.reporter.Reporter._writeln
|   Warning: @param for unknown parameter "*args"
```

When I change it back to '@param args' the error goes away and the docs are properly generated with epydoc. pydoctor might already understand this but I think it's a good idea to make it work for both epydoc and pydoctor.

I've changed the 'args' param to '*args' in r25096 and opened a ticket ([#3501](https://github.com/twisted/twisted/issues/3501)) that proposes to fix all of these occurrences in the codebase.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [25097]) Merge proper-epytext-2375: Fix some incorrect epytext markup in the docstrings.

Author: thijs
Reviewer: exarkun
Fixes: [#2375](https://github.com/twisted/twisted/issues/2375)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
